src/VirtoCommerce.NotificationsModule.Data/Services/NotificationSearchService.cs

```diff
@@ -25,68 +25,86 @@ public NotificationSearchService(Func<INotificationRepository> repositoryFactory
+                    var notificationIds = await query.OrderBySortInfos(sortInfos).ThenBy(x => x.Id)
                                                      .Select(x => x.Id)
                                                      .Skip(criteria.Skip).Take(criteria.Take)
                                                      .ToArrayAsync();
                     var unorderedResults = await _notificationService.GetByIdsAsync(notificationIds, criteria.ResponseGroup);
-                    result.Results = unorderedResults.OrderBy(x => Array.IndexOf(notificationIds, x.Id)).ToList();
+                    result = unorderedResults.OrderBy(x => Array.IndexOf(notificationIds, x.Id)).ToArray();
                 }
             }
-            tmpSkip = Math.Min(result.TotalCount, criteria.Skip);
-            tmpTake = Math.Min(criteria.Take, Math.Max(0, result.TotalCount - criteria.Skip));
-
-            criteria.Skip -= tmpSkip;
-            criteria.Take -= tmpTake;
-
-            if (criteria.Take > 0)
-            {
-                var transientNotificationsQuery = AbstractTypeFactory<Notification>.AllTypeInfos.Select(x =>
-                {
-                    return AbstractTypeFactory<Notification>.TryCreateInstance(x.Type.Name);
-                })
-                                                                              .OfType<Notification>().AsQueryable();
-                if (!string.IsNullOrEmpty(criteria.NotificationType))
-                {
-                    transientNotificationsQuery = transientNotificationsQuery.Where(x => x.Type.EqualsInvariant(criteria.NotificationType));
-                }
-
-                var allPersistentProvidersTypes = result.Results.Select(x => x.GetType()).Distinct();
-                transientNotificationsQuery = transientNotificationsQuery.Where(x => !allPersistentProvidersTypes.Contains(x.GetType()));
 
-                transientNotificationsQuery = transientNotificationsQuery.Where(x => !x.Kind.EqualsInvariant(x.Type));
-
-                result.TotalCount += transientNotificationsQuery.Count();
-                var transientNotifications = transientNotificationsQuery.Skip(criteria.Skip).Take(criteria.Take).ToList();
-                foreach(var transientNotification in transientNotifications)
-                {
-                    transientNotification.ReduceDetails(criteria.ResponseGroup);
-                }
-                result.Results = result.Results.Concat(transientNotifications)
-                    .AsQueryable()
-                    .OrderBySortInfos(sortInfos).ToList();
-            }
             return result;
         }
 
         protected virtual IQueryable<NotificationEntity> BuildQuery(INotificationRepository repository, NotificationSearchCriteria criteria, IEnumerable<SortInfo> sortInfos)
         {
             var query = repository.Notifications;
 
-            if (!string.IsNullOrEmpty(criteria.NotificationType))
+            if (!criteria.NotificationTypes.IsNullOrEmpty())
```

Comment:
The notification "Alias" can be checked here

Reply:
then need to add the property to DB(add migration and etc..), is it OK?

Reply:
look at the code, make some refactor